File: src/lib/scrapers/reddit_scraper.py
"""
Reddit scraper using public JSON API (no authentication required)
"""

import requests
import json
import sys
import time
from datetime import datetime

def scrape_reddit_no_auth(bike1, bike2):
    """
    Scrape Reddit r/IndianBikes using public JSON API
    
    Args:
        bike1: First bike name
        bike2: Second bike name
    
    Returns:
        JSON string with posts and comments
    """
    
    results = {
        "bike1": {"name": bike1, "posts": []},
        "bike2": {"name": bike2, "posts": []},
        "metadata": {
            "scraped_at": datetime.now().isoformat(),
            "source": "reddit",
            "subreddit": "IndianBikes",
            "total_posts": 0,
            "total_comments": 0
        }
    }
    
    headers = {
        'User-Agent': 'VehicleResearchBot/1.0 (Educational Research)'
    }
    
    for bike_key, bike_name in [("bike1", bike1), ("bike2", bike2)]:
        # Reddit's public JSON search endpoint
        search_url = "https://www.reddit.com/r/IndianBikes/search.json"
        params = {
            'q': bike_name,
            'restrict_sr': 'on',  # Search only in r/IndianBikes
            'limit': 10,  # Reduced from 20 to 10 for speed
            'sort': 'relevance',
            'type': 'link'
        }
        
        try:
            # Search for posts
            print(f"Searching Reddit for: {bike_name}", file=sys.stderr)
            response = requests.get(search_url, headers=headers, params=params, timeout=15)
            response.raise_for_status()
            data = response.json()
            
            posts_found = 0
            
            for post in data['data']['children']:
                post_data = post['data']
                
                post_obj = {
                    "id": post_data['id'],
                    "title": post_data['title'],
                    "author": post_data.get('author', '[deleted]'),
                    "score": post_data['score'],
                    "url": f"https://reddit.com{post_data['permalink']}",
                    "created_utc": post_data['created_utc'],
                    "num_comments": post_data['num_comments'],
                    "selftext": post_data.get('selftext', '')[:500],
                    "comments": []
                }
                
                # Get top comments (via JSON)
                comments_url = f"https://www.reddit.com{post_data['permalink']}.json"
                
                try:
                    # Rate limiting: wait 1 second between requests
                    time.sleep(1)
                    
                    print(f"Fetching comments for post: {post_obj['title'][:50]}...", file=sys.stderr)
                    comments_response = requests.get(comments_url, headers=headers, timeout=10)
                    comments_response.raise_for_status()
                    comments_data = comments_response.json()
                    
                    # Comments are in the second element of the response
                    if len(comments_data) > 1 and 'data' in comments_data[1]:
                        comment_listing = comments_data[1]['data']['children']
                        
                        for comment in comment_listing[:5]:  # Top 5 comments (faster)
                            if comment.get('kind') == 't1':  # Ensure it's a comment
                                c_data = comment['data']
                                
                                # Skip deleted/removed comments
                                if c_data.get('author') in ['[deleted]', 'AutoModerator']:
                                    continue
                                
                                comment_obj = {
                                    "id": c_data['id'],
                                    "author": c_data.get('author', '[deleted]'),
                                    "body": c_data.get('body', '')[:300],
                                    "score": c_data['score'],
                                    "created_utc": c_data['created_utc']
                                }
                                
                                post_obj["comments"].append(comment_obj)
                                results["metadata"]["total_comments"] += 1
                        
                        print(f"  → Found {len(post_obj['comments'])} comments", file=sys.stderr)
                        
                except Exception as e:
                    print(f"Warning: Could not fetch comments for post {post_obj['id']}: {e}", file=sys.stderr)
                    # Continue even if comments fail
                
                results[bike_key]["posts"].append(post_obj)
                results["metadata"]["total_posts"] += 1
                posts_found += 1
            
            print(f"✓ Found {posts_found} posts for {bike_name}", file=sys.stderr)
            
        except requests.exceptions.RequestException as e:
            print(f"Error scraping Reddit for {bike_name}: {e}", file=sys.stderr)
        except Exception as e:
            print(f"Unexpected error for {bike_name}: {e}", file=sys.stderr)
    
    return json.dumps(results, indent=2)

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print(json.dumps({"error": "Usage: python reddit_scraper.py 'bike1' 'bike2'"}))
        sys.exit(1)
    
    bike1 = sys.argv[1]
    bike2 = sys.argv[2]
    
    try:
        result = scrape_reddit_no_auth(bike1, bike2)
        print(result)
    except Exception as e:
        print(json.dumps({"error": str(e)}), file=sys.stderr)
        sys.exit(1)

File: src/lib/scrapers/reddit-scraper.ts
// Node.js Reddit scraper - works on Vercel!
// Replacement for Python scraper using Reddit's public JSON API

interface RedditPost {
  title: string;
  url: string;
  selftext: string;
  author: string;
  created_utc: number;
  num_comments: number;
  score: number;
  permalink: string;
  comments: RedditComment[];
}

interface RedditComment {
  author: string;
  body: string;
  score: number;
  created_utc: number;
}

interface BikeData {
  name: string;
  posts: RedditPost[];
}

interface RedditScrapingResult {
  bike1: BikeData;
  bike2: BikeData;
  metadata: {
    scraped_at: string;
    source: string;
    subreddit: string;
    total_posts: number;
    total_comments: number;
  };
}

/**
 * Scrapes Reddit for a single bike
 */
async function scrapeRedditForBike(bikeName: string): Promise<RedditPost[]> {
  const subreddit = 'IndianBikes';
  const limit = 10; // Number of posts to fetch per bike
  
  try {
    // Step 1: Search for posts
    const searchUrl = `https://www.reddit.com/r/${subreddit}/search.json`;
    const searchParams = new URLSearchParams({
      q: bikeName,
      restrict_sr: 'on',
      limit: limit.toString(),
      sort: 'relevance',
      t: 'all'
    });

    console.log(`[Reddit] Searching for: ${bikeName}`);
    
    const searchResponse = await fetch(`${searchUrl}?${searchParams}`, {
      headers: {
        'User-Agent': 'BikeDekho-Research/1.0 (Educational Project)',
        'Accept': 'application/json'
      }
    });

    if (!searchResponse.ok) {
      throw new Error(`Reddit API error: ${searchResponse.status}`);
    }

    const searchData = await searchResponse.json();
    const posts = searchData.data?.children || [];

    console.log(`[Reddit] Found ${posts.length} posts for ${bikeName}`);

    // Step 2: Fetch comments for each post
    const postsWithComments: RedditPost[] = [];
    
    for (const post of posts) {
      const postData = post.data;
      
      // Fetch top 5 comments for this post
      const comments = await fetchPostComments(postData.permalink, 5);
      
      postsWithComments.push({
        title: postData.title,
        url: postData.url,
        selftext: postData.selftext || '',
        author: postData.author,
        created_utc: postData.created_utc,
        num_comments: postData.num_comments,
        score: postData.score,
        permalink: postData.permalink,
        comments: comments
      });

      // Small delay to be respectful to Reddit's servers
      await delay(100);
    }

    return postsWithComments;

  } catch (error) {
    console.error(`[Reddit] Error scraping for ${bikeName}:`, error);
    throw error;
  }
}

/**
 * Fetches top comments for a specific post
 */
async function fetchPostComments(permalink: string, limit: number = 5): Promise<RedditComment[]> {
  try {
    const commentsUrl = `https://www.reddit.com${permalink}.json`;
    
    const response = await fetch(commentsUrl, {
      headers: {
        'User-Agent': 'BikeDekho-Research/1.0 (Educational Project)',
        'Accept': 'application/json'
      }
    });

    if (!response.ok) {
      console.warn(`[Reddit] Failed to fetch comments for ${permalink}`);
      return [];
    }

    const data = await response.json();
    
    // Reddit returns [post_data, comments_data]
    const commentsData = data[1]?.data?.children || [];
    
    const comments: RedditComment[] = [];
    
    for (const comment of commentsData.slice(0, limit)) {
      const commentData = comment.data;
      
      // Skip deleted/removed comments and "more comments" markers
      if (commentData.body && 
          commentData.body !== '[deleted]' && 
          commentData.body !== '[removed]' &&
          commentData.kind !== 'more') {
        comments.push({
          author: commentData.author || 'unknown',
          body: commentData.body,
          score: commentData.score || 0,
          created_utc: commentData.created_utc || 0
        });
      }
    }

    return comments;

  } catch (error) {
    console.warn(`[Reddit] Error fetching comments for ${permalink}:`, error);
    return [];
  }
}

/**
 * Main function: Scrapes Reddit for two bikes
 * Returns data in same format as Python scraper for compatibility
 */
export async function scrapeRedditForComparison(
  bike1: string, 
  bike2: string
): Promise<RedditScrapingResult> {
  console.log(`[Reddit] Starting scraping for ${bike1} vs ${bike2}`);
  
  // Scrape both bikes in parallel for speed
  const [bike1Posts, bike2Posts] = await Promise.all([
    scrapeRedditForBike(bike1),
    scrapeRedditForBike(bike2)
  ]);

  // Calculate totals
  const totalPosts = bike1Posts.length + bike2Posts.length;
  const totalComments = 
    bike1Posts.reduce((sum, post) => sum + post.comments.length, 0) +
    bike2Posts.reduce((sum, post) => sum + post.comments.length, 0);

  console.log(`[Reddit] Scraping complete: ${totalPosts} posts, ${totalComments} comments`);

  // Return in same format as Python scraper
  return {
    bike1: {
      name: bike1,
      posts: bike1Posts
    },
    bike2: {
      name: bike2,
      posts: bike2Posts
    },
    metadata: {
      scraped_at: new Date().toISOString(),
      source: 'reddit',
      subreddit: 'IndianBikes',
      total_posts: totalPosts,
      total_comments: totalComments
    }
  };
}

/**
 * Helper function to add delay
 */
function delay(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

